### PR TITLE
[core, node, darwin, qt] Remove support for paint classes

### LINKED ISSUE #2875: Remove style class support
It is time to mark the style class APIs as deprecated in iOS and Android.

Mapbox Studio does not support creating styles with classes in them. None of the bundled styles uses them. Neither the iOS nor Android test apps/Sirius touch these APIs.

As far as I know they could be completely broken and we would not even notice.

We should first update the documentation to mark them as deprecated.

Eventually we can remove them entirely.

This does not require any changes to core GL at this stage. Basically will only affect the higher level OS specific wrappers.

This will allow us to close:
https://github.com/mapbox/mapbox-gl-native/issues/794
https://github.com/mapbox/mapbox-gl-native/issues/1744
https://github.com/mapbox/mapbox-gl-native/issues/2077

@mikemorris Does node still need to support style classes?

@1ec5 Are there any open iOS tickets this will close?

### PR DESCRIPTION
**Pull request by jfirebaugh**
We've long talked about doing this, and with the implementation of async rendering, which will unlock [smart set style](https://github.com/mapbox/mapbox-gl-native/issues/7893), I think it's time.

Fixes #2875.

### CODE REVIEW

File: platform/darwin/src/MGLStyle.mm [hunk 1]
```diff
@@ -469,60 +469,32 @@ - (void)insertLayer:(MGLStyleLayer *)layer aboveLayer:(MGLStyleLayer *)sibling {
...
 - (NSUInteger)countOfStyleClasses {
-    const auto &classes = self.mapView.mbglMap->getClasses();
-    return classes.size();
+    return 0;
 }
 
 - (BOOL)hasStyleClass:(NSString *)styleClass
 {
-    return styleClass && self.mapView.mbglMap->hasClass([styleClass UTF8String]);
+    return false;
```
Comment (1ec5) [inline platform/darwin/src/MGLStyle.mm]: The negative version of `BOOL` is `NO`, not `false`.

File: platform/darwin/src/MGLStyle.mm [hunk 2]
```diff
@@ -469,60 +469,32 @@ - (void)insertLayer:(MGLStyleLayer *)layer aboveLayer:(MGLStyleLayer *)sibling {
...
-    const std::vector<std::string> &appliedClasses = self.mapView.mbglMap->getClasses();
-
-    NSMutableArray *returnArray = [NSMutableArray arrayWithCapacity:appliedClasses.size()];
-
-    for (auto appliedClass : appliedClasses) {
-       [returnArray addObject:@(appliedClass.c_str())];
-    }
-
-    return returnArray;
+    return [NSMutableArray arrayWithCapacity:0];
```
Comment (1ec5) [inline platform/darwin/src/MGLStyle.mm]: Use `@[]` for an empty array.

File: platform/darwin/src/MGLStyle.h [hunk 3]
```diff
@@ -483,40 +483,24 @@ MGL_EXPORT
...
+- (BOOL)hasStyleClass:(NSString *)styleClass __attribute__((deprecated("This method is non-functional.")));
 
 /**
- Activates the style class with the given identifier.
-
- @param styleClass The style class to activate.
+ Support for style classes has been removed. This method is a no-op.
  */
-- (void)addStyleClass:(NSString *)styleClass __attribute__((deprecated("This method will be removed in a future release.")));
+- (void)addStyleClass:(NSString *)styleClass __attribute__((deprecated("This method is non-functional.")));
```
Comment (1ec5) [inline platform/darwin/src/MGLStyle.h]: Normally deprecation messages say what to use instead, or they say, “There is no replacement.” In this case, the replacement would be setting properties on style layers individually, right?

File: platform/qt/src/qmapboxgl.cpp [hunk 4]
```diff
@@ -752,66 +742,55 @@ void QMapboxGL::setGestureInProgress(bool progress)
...
-    This was removed from the \l {https://www.mapbox.com/mapbox-gl-style-spec/#layer-paint.*}
-    {Mapbox style specification} and should no longer be used.
+    Paint classes were removed from the \l {https://www.mapbox.com/mapbox-gl-style-spec/#layer-paint.*}
+    {Mapbox style specification}; this method no longer has any effect.
 
     \deprecated
-    \sa removeClass()
 */
-void QMapboxGL::addClass(const QString &className)
+void QMapboxGL::addClass(const QString &)
```
Comment (brunoabinader) [inline platform/qt/src/qmapboxgl.cpp]: We don't need to maintain binary compatibility on the QMapboxGL interface - we use it internally and expose a separate API on the QtLocation plugin - thus it is safe to assume we can remove all class-handling function implementations from QMapboxGL.

File: platform/qt/src/qmapboxgl.cpp [hunk 5]
```diff
@@ -752,66 +742,55 @@ void QMapboxGL::setGestureInProgress(bool progress)
...
 /*!
-    Removes a \a className.
+    Paint classes were removed from the \l {https://www.mapbox.com/mapbox-gl-style-spec/#layer-paint.*}
+    {Mapbox style specification}; this method no longer has any effect.
 
     \deprecated
-    \sa addClass()
 */
-void QMapboxGL::removeClass(const QString &className)
+void QMapboxGL::removeClass(const QString &)
```
Comment (brunoabinader) [inline platform/qt/src/qmapboxgl.cpp]: Ditto.

File: platform/qt/src/qmapboxgl.cpp [hunk 6]
```diff
@@ -752,66 +742,55 @@ void QMapboxGL::setGestureInProgress(bool progress)
...
 /*!
-    Returns true when \a className is active, false otherwise.
+    Paint classes were removed from the \l {https://www.mapbox.com/mapbox-gl-style-spec/#layer-paint.*}
+    {Mapbox style specification}; this method always returns false.
 
     \deprecated
-    \sa addClass()
 */
-bool QMapboxGL::hasClass(const QString &className) const
+bool QMapboxGL::hasClass(const QString &) const
```
Comment (brunoabinader) [inline platform/qt/src/qmapboxgl.cpp]: Ditto.

File: platform/qt/src/qmapboxgl.cpp [hunk 7]
```diff
@@ -752,66 +742,55 @@ void QMapboxGL::setGestureInProgress(bool progress)
...
 /*!
-    Bulk adds a list of \a classNames.
+    Paint classes were removed from the \l {https://www.mapbox.com/mapbox-gl-style-spec/#layer-paint.*}
+    {Mapbox style specification}; this method no longer has any effect.
 
     \deprecated
-    \sa addClass()
 */
-void QMapboxGL::setClasses(const QStringList &classNames)
+void QMapboxGL::setClasses(const QStringList &)
```
Comment (brunoabinader) [inline platform/qt/src/qmapboxgl.cpp]: Ditto.

File: platform/qt/src/qmapboxgl.cpp [hunk 8]
```diff
@@ -752,66 +742,55 @@ void QMapboxGL::setGestureInProgress(bool progress)
...
 
 /*!
-    Returns a list of active classes.
+    Paint classes were removed from the \l {https://www.mapbox.com/mapbox-gl-style-spec/#layer-paint.*}
+    {Mapbox style specification}; this method always returns an empty list.
 
     \deprecated
-    \sa setClasses()
 */
 QStringList QMapboxGL::getClasses() const
```
Comment (brunoabinader) [inline platform/qt/src/qmapboxgl.cpp]: Ditto.

File: platform/qt/src/qmapboxgl.cpp [hunk 9]
```diff
@@ -1006,7 +984,7 @@ void QMapboxGL::setLayoutProperty(const QString& layer, const QString& property,
         map->setPaintProperty("route","line-dasharray", lineDashArray);
     \endcode
 */
-void QMapboxGL::setPaintProperty(const QString& layer, const QString& property, const QVariant& value, const QString& styleClass)
+void QMapboxGL::setPaintProperty(const QString& layer, const QString& property, const QVariant& value, const QString&)
```
Comment (brunoabinader) [inline platform/qt/src/qmapboxgl.cpp]: For the same reasons previously mentioned, we can remove the final argument entirely.

**Comment by jfirebaugh** [comment]
Great, thanks @brunoabinader, I will remove it.

**Pull request merged**